contracts/ecash/src/contract.rs

```diff
@@ -0,0 +1,430 @@
+// Copyright 2024 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: Apache-2.0
+
+use cosmwasm_std::{
+    to_binary, BankMsg, Coin, CosmosMsg, Event, Order, Reply, Response, StdError, StdResult,
+    SubMsg, WasmMsg,
+};
+use cw3::ProposalResponse;
+use cw4::Cw4Contract;
+use cw_controllers::Admin;
+use cw_storage_plus::{Bound, IndexedMap, Item};
+use nym_ecash_contract_common::blacklist::{
+    BlacklistProposal, BlacklistedAccount, BlacklistedAccountResponse,
+    PagedBlacklistedAccountResponse,
+};
+use nym_ecash_contract_common::events::{
+    BLACKLIST_PROPOSAL_ID, BLACKLIST_PROPOSAL_REPLY_ID, DEPOSITED_FUNDS_EVENT_TYPE,
+    DEPOSIT_ENCRYPTION_KEY, DEPOSIT_IDENTITY_KEY, DEPOSIT_INFO, DEPOSIT_VALUE,
+};
+use nym_ecash_contract_common::msg::ExecuteMsg;
+use nym_multisig_contract_common::msg::ExecuteMsg as MultisigExecuteMsg;
+use nym_multisig_contract_common::msg::QueryMsg as MultisigQueryMsg;
+use sylvia::types::{ExecCtx, InstantiateCtx, QueryCtx, ReplyCtx};
+use sylvia::{contract, entry_points};
+
+use crate::errors::ContractError;
+use crate::state::Config;
+use crate::storage::{
+    self, BlacklistIndex, BlacklistProposalIndex, SpendCredentialIndex,
+    BLACKLIST_PAGE_DEFAULT_LIMIT, BLACKLIST_PAGE_MAX_LIMIT, SPEND_CREDENTIAL_PAGE_DEFAULT_LIMIT,
+    SPEND_CREDENTIAL_PAGE_MAX_LIMIT,
+};
+use nym_ecash_contract_common::events::{TICKET_BOOK_VALUE, TICKET_VALUE};
+use nym_ecash_contract_common::spend_credential::{
+    EcashSpentCredential, EcashSpentCredentialResponse, PagedEcashSpentCredentialResponse,
+};
+
+pub struct NymEcashContract<'a> {
+    pub(crate) admin: Admin<'a>,
+    pub(crate) config: Item<'a, Config>,
+    pub(crate) spent_credentials:
+        IndexedMap<'a, &'a str, EcashSpentCredential, SpendCredentialIndex<'a>>,
+    pub(crate) blacklist: IndexedMap<'a, &'a str, BlacklistedAccount, BlacklistIndex<'a>>,
+    pub(crate) blacklist_proposals:
+        IndexedMap<'a, &'a str, BlacklistProposal, BlacklistProposalIndex<'a>>,
+}
+
+#[entry_points]
+#[contract]
+#[error(ContractError)]
+impl NymEcashContract<'_> {
```

Comment:
add the migrate function that updates the cw2 info (like we have in mixnet/vesting/etc)

Reply:
Added in 34c5c6a6201faa79e00e9b68fe18aef738e65d26